source/extensions/filters/network/dubbo_proxy/active_message.h

```diff
@@ -27,128 +27,182 @@ namespace DubboProxy {
+class ActiveResponseDecoder : public ResponseDecoderCallbacks,
+                              public StreamHandler,
+                              Logger::Loggable<Logger::Id::dubbo> {
 public:
-  ResponseDecoder(Buffer::Instance& buffer, DubboFilterStats& stats,
-                  Network::Connection& connection, Deserializer& deserializer, Protocol& protocol);
-  ~ResponseDecoder() override = default;
+  ActiveResponseDecoder(ActiveMessage& parent, DubboFilterStats& stats,
+                        Network::Connection& connection, ProtocolPtr&& protocol);
+  ~ActiveResponseDecoder() override = default;
 
-  bool onData(Buffer::Instance& data);
+  DubboFilters::UpstreamResponseStatus onData(Buffer::Instance& data);
 
-  // DecoderEventHandler
-  Network::FilterStatus transportBegin() override;
-  Network::FilterStatus transportEnd() override;
-  Network::FilterStatus messageBegin(MessageType type, int64_t message_id,
-                                     SerializationType serialization_type) override;
-  Network::FilterStatus messageEnd(MessageMetadataSharedPtr metadata) override;
+  // StreamHandler
+  void onStreamDecoded(MessageMetadataSharedPtr metadata, ContextSharedPtr ctx) override;
 
-  // DecoderCallbacks
-  DecoderEventHandler* newDecoderEventHandler() override;
+  // ResponseDecoderCallbacks
+  StreamHandler& newStream() override { return *this; }
+  void onHeartbeat(MessageMetadataSharedPtr) override { NOT_IMPLEMENTED_GCOVR_EXCL_LINE; }
 
   uint64_t requestId() const { return metadata_ ? metadata_->request_id() : 0; }
 
 private:
-  Buffer::Instance& response_buffer_;
+  FilterStatus applyMessageEncodedFilters(MessageMetadataSharedPtr metadata, ContextSharedPtr ctx);
+
+  ActiveMessage& parent_;
   DubboFilterStats& stats_;
   Network::Connection& response_connection_;
-  DecoderPtr decoder_;
+  ProtocolPtr protocol_;
+  ResponseDecoderPtr decoder_;
   MessageMetadataSharedPtr metadata_;
   bool complete_ : 1;
+  DubboFilters::UpstreamResponseStatus response_status_;
 };
 
-typedef std::unique_ptr<ResponseDecoder> ResponseDecoderPtr;
+typedef std::unique_ptr<ActiveResponseDecoder> ActiveResponseDecoderPtr;
+
+class ActiveMessageFilterBase : public virtual DubboFilters::FilterCallbacksBase {
+public:
+  ActiveMessageFilterBase(ActiveMessage& parent, bool dual_filter)
+      : parent_(parent), dual_filter_(dual_filter) {}
+  ~ActiveMessageFilterBase() override = default;
+
+  // DubboFilters::FilterCallbacksBase
+  uint64_t requestId() const override;
+  uint64_t streamId() const override;
+  const Network::Connection* connection() const override;
+  DubboProxy::Router::RouteConstSharedPtr route() override;
+  SerializationType serializationType() const override;
+  ProtocolType protocolType() const override;
+  StreamInfo::StreamInfo& streamInfo() override;
+  Event::Dispatcher& dispatcher() override;
+  void resetStream() override;
+
+protected:
+  ActiveMessage& parent_;
+  const bool dual_filter_ : 1;
```

Comment:
Where use?

Reply:
ActiveMessageDecoderFilter and ActiveMessagerEncodeFilter will use dual-filter.